docs for btc light client (#424)
babylonchain/babylon: x/btclightclient/README.md

# BTC light client

> 

Reviewer: Perhaps we can start by saying "The BTC light client module is essentially a BTC light client that maintains the canonical header chain of Bitcoin."

> Babylon chain needs to know about different events happening on Bitcoin chain
> To make it possible in secure way, Babylon needs to known what is the current
> state of Bitcoin chain i.e what is best chain recognized by Bitcoin network.

Reviewer 2: ```suggestion
Babylon chain needs to know about different events happening on Bitcoin chain.
To make it possible in a secure way, Babylon needs to know the current
state of Bitcoin chain i.e, what is the canonical chain of the Bitcoin network.
```

## Table of contents

- [Table of contents](#table-of-contents)

> - [Concepts](#concepts)

Reviewer 2: The section title "Concepts" seems missing

  - [Problem statement](#problem-statement)
  - [Babylon's BTC light client design](#babylons-btc-light-client-design)
- [States](#states)
  - [Parameters](#parameters)
  - [Headers storage](#headers-storage)
  - [HashToHeight storage](#hashtoheight-storage)
- [Messages](#messages)
  - [MsgInsertHeaders](#msginsertheaders)
  - [MsgUpdateParams](#msgupdateparams)
- [Hooks](#hooks)
  - [Hooks exposed by BTC light client](#hooks-exposed-by-btc-light-client)
- [Events](#events)

### Problem statement

In the Babylon chain, the are few events which needs to be validated against

> current state of BTC chain.

Reviewer 2: The Babylon chain needs to learn and validate a number of events happened on the Bitcoin chain.

> Those events are:

Reviewer: Maybe this line can be inserted at the end of the previous paragragh

> 1. Bitcoin Timestamping protocol requires to report back checkpoints which were

Reviewer: Maybe for each numbered point, we can start with an event name. For example, here we can use `New checkpoint events`.

> 1. Bitcoin Timestamping protocol requires to report back checkpoints which were
> send to BTC chain. To do it securely, each checkpoint must be reported back along
> with BTC inclusion proof of transaction which carried given checkpoint.

Reviewer 2: ```suggestion
1. Bitcoin Timestamping protocol requires checkpoints on Bitcoin to be reported back to Babylon. To do it securely, each checkpoint must be reported back along
with the inclusion proof of transactions which carry this checkpoint.
```

> 2. Bitcoin staking protocol requires that staking transaction reported to
> Babylon must have proper depth in BTC chain and also must ba accompanied by BTC
> inclusion proof.

Reviewer 2: ```suggestion
2. Bitcoin Staking protocol requires staking transactions to be deep enough in the Bitcoin chain. Thus the staking transactions also must be accompanied by the
inclusion proof.
```

> To properly validate those inclusion proofs, the Babylon chain needs to know what
> is current state of BTC chain i.e what is current canonical chain recognized by BTC

Reviewer 2: ```suggestion
To properly validate those inclusion proofs, the Babylon chain needs to know the current state of BTC chain i.e, what is current canonical chain recognized by BTC
```

network.

### Babylon's BTC light client design

> 

Reviewer 2: Maybe a sentence that "Babylon maintains a BTC light client so that it can verify the inclusion of Bitcoin transactions." in order to answer the above problem statement

> In high level overview, Babylon BTC light client starts from some initial BTC
> header existing on BTC network and allow to extend this header by applying the
> same rules as normal BTC node.

Reviewer: ```suggestion
In a high-level overview, Babylon BTC light client starts from some initial BTC
header existing on the BTC network and allows extending this header by applying the
same rules as a normal BTC node.
```

> Initial BTC header must:

Reviewer: Should we mention that is a genesis parameter of the btc lc module? And in the module, we use the term `base header`. Perhaps we should keep it consistent here.

- be deep enough in BTC chain so that it will never be reverted by BTC network.

> - be at height which is at BTC difficulty adjustment boundary. This required, to

Reviewer 2: ```suggestion
- be at height at BTC difficulty adjustment boundary. This is required, to
```
Also a link to the "difficulty adjustment boundary" design in Bitcoin could be helpful

properly validate all future difficulty adjustments.

The Babylon BTC light client module stores only BTC canonical header chain and

> does not store the headers which end of forks. The BTC canonical chain can only

Reviewer 2: ```suggestion
does not store the headers on the forks. The BTC canonical chain can only
```

> be extended by processing valid [MsgInsertHeaders](#msginsertheaders) message.

Reviewer 2: ```suggestion
be extended by processing valid [MsgInsertHeaders](#msginsertheaders) messages.
```

If the better fork is encountered:

> 1. current chain is rollbacked up to header which is parent of received fork

Reviewer 2: ```suggestion
1. current chain is rolled back to the parent of the received fork
```

> 2. chain is extend with new header from the fork

Reviewer 2: ```suggestion
2. chain is extend with new headers from the fork
```

## States

The BTC light client module maintains the following KV stores.

### Parameters

The [parameter storage](./keeper/params.go) maintains the BTC light client module's
parameters. The BTC light client module's parameters are represented as a `Params`
[object](../../proto/babylon/btclightclient/v1/params.proto) defined as follows:

```protobuf
// Params defines the parameters for the module.
message Params {
  option (gogoproto.equal) = true;

  // List of addresses which are allowed to insert headers to btc light client
  // if the list is empty, any address can insert headers
  repeated string insert_headers_allow_list = 1;
}
```

In nutshell, `insert_headers_allow_list` makes it possible to set up
restrictions about who is able to update BTC light client module state.

If `insert_headers_allow_list` is not empty, only addresses in the list can send
`MsgInsertHeaders` message.

### Headers storage

> The [Headers storage](./keeper/state.go) maintains information about each header.

Reviewer 2: ```suggestion
The [Headers storage](./keeper/state.go) maintains all headers on the canonical chain of Bitcoin.
```

The key is the header height in BTC chain, and the value is an `BTCHeaderInfo`
[object](../../proto/babylon/btclightclient/v1/btclightclient.proto)
which contains BTC header along with some metadata.

```protobuf
// BTCHeaderInfo is a structure that contains all relevant information about a
// BTC header
//  - Full header bytes
//  - Header hash for easy retrieval
//  - Height of the header in the BTC chain
//  - Total work spent on the header. This is the sum of the work corresponding
//  to the header Bits field
//    and the total work of the header.
message BTCHeaderInfo {
  bytes header = 1
      [ (gogoproto.customtype) =
            "github.com/babylonchain/babylon/types.BTCHeaderBytes" ];
  bytes hash = 2
      [ (gogoproto.customtype) =
            "github.com/babylonchain/babylon/types.BTCHeaderHashBytes" ];
  uint64 height = 3;
  bytes work = 4
      [ (gogoproto.customtype) = "cosmossdk.io/math.Uint" ];
}
```

### HashToHeight storage

> The [HashToHeight storage](./keeper/state.go) maintains index in which key is

Reviewer 2: ```suggestion
The [HashToHeight storage](./keeper/state.go) maintains an index in which key is
```

BTC header hash and value is BTC header height.

> This index enables efficient lookup of BTC headers by their hash. This useful

Reviewer 2: ```suggestion
This index enables efficient lookup of BTC headers by their hash. This is useful
```

> in many situations, but mostly when receiving potential chain extension which
> does not points to current BTC chain tip.

Reviewer 2: ```suggestion
in many situations, notably when receiving potential chain extension which
does not point to the current BTC chain tip.
```

## Messages

### MsgInsertHeaders

`MsgInsertHeaders` is the main message processed by BTC light client module.

> Its purpose is to update the state of the BTC chain as viewed by Babylon chain.

Reviewer 2: How about we provide a link to the implementation of the message handler?

```proto
message MsgInsertHeaders {
  option (cosmos.msg.v1.signer) = "signer";

  string signer = 1;
  repeated bytes headers = 2
      [ (gogoproto.customtype) =
            "github.com/babylonchain/babylon/types.BTCHeaderBytes" ];
}
```

> The message contains a list of BTC headers encoded in Bitcoin format.

Reviewer 2: This sentence can be moved to the front of the protobuf definiton

> Rules applied when processing the message:

Reviewer 2: ```suggestion
Upon a `MsgInsertHeaders` message, a Babylon node applies the following verification rules:
```
Also, since these verification rules are not invented by us, we could provide some links about these verification rules in Bitcoin.

Author: sure I will point to https://en.bitcoin.it/wiki/Protocol_rules#.22block.22_messages, and add annotation that we use only subset of those, related to headers.

- `headers` list must not be empty

> - headers in the list must be connected by parent-child relationships i.e

Reviewer 2: ```suggestion
- headers in the list must be connected by parent-child relationships i.e.,
```

header at position `i + 1`, must have its `PrevBlock` field set to header `i`
hash
- first header of the list must point to the header already maintained by BTC
light client module
- each header must be correctly encoded
- each header in the list must have valid proof of work and difficulty
- each header in the list must have `Timestamp` which happened after median
of last 11 ancestors
- if the first header of the list does not point to the current tip of the
chain maintained by BTC light client, it means that message contains a fork. For

> fork to be valid, forked chain must be better than current chain maintained by

Reviewer 2: Word "better" might need to be defined somewhere. For example "a chain is better than another fork if the total work of proof-of-work solutions in this chain is more than that in the fork". Also this is pretty much the core of Bitcoin's fork choice so we could find some references for it

Reviewer 3: Does "better" just means "longer" here? Or are there other conditions?

Reviewer 2: Not really, a chain is better than another chain if it has [more work](https://github.com/babylonchain/babylon/blob/30d1644d00e4676f59eb9ebe534281742083c56a/x/btclightclient/types/btc_light_client.go#L392-L395) (in terms of the proof of work mechanism in Bitcoin) on it. 

Although the fork choice rule in Bitcoin is usually referred to as "longest-chain rule", the implementation is a bit different. I suppose this is because in most cases, a longer chain has more work on it in Bitcoin.

Author: so difficulty is adjusted roughly every 2weeks and during this time difficulty is constant, this leads to situtaion that in most cases this is longest chain rule.

Although I imagine at difficulty adjustment boundaries it can happen that chain of same length has more work, for example:
lets say height `x` is difficulty adjustment boundary, if two miners mined block at this height then this blocks will have different adjusted difficulty (as adjusted difficulty is function of time and most probably blocks were mined at different times). The network will accept chain with block at height `x` ,  which has higher calculated difficulty. S

> BTC light client. The fork is better when its total work is larger that the work

Reviewer 3: It's clarified here. Perhaps just replacing "better" by "longer" / "larger" is simpler / clearer.

of current chain

> All those rules are the same rules which are applied by BTC nodes when receiving

Reviewer 3: 👍🏼

headers from the BTC network.

Processing of the message is atomic, so even if one header in the list is
invalid, the state of the BTC light client module won't be updated.

> In case of receiving valid chain extension i.e first header of the `headers` list
> points to current BTC chain tip, the chain maintained by BTC light client module

Reviewer 2: ```suggestion
In case of receiving valid chain extension, the chain maintained by BTC light client module
```
We have defined "valid chain extension" before so no need to mention it again

does not point to to current BTC chain tip and fork total work is larger than
current chain BTC chain total work, chain maintained by BTC light client will
be roll backed to header which is a fork header, and then it will be
extended with headers received in `headers`.

### MsgUpdateParams

The `MsgUpdateParams` message is used for updating the module parameters for the
BTC light client module. It can only be executed via a governance proposal.

```protobuf
// MsgUpdateParams defines a message for updating btc light client module parameters.
message MsgUpdateParams {
  option (cosmos.msg.v1.signer) = "authority";

  // authority is the address of the governance account.
  // just FYI: cosmos.AddressString marks that this field should use type alias
  // for AddressString instead of string, but the functionality is not yet implemented
  // in cosmos-proto
  string authority = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];

  // params defines the btc light client parameters to update.
  //
  // NOTE: All parameters must be supplied.
  Params params = 2 [(gogoproto.nullable) = false];
}
```

## Hooks

The BTC light client module exposes a set of hooks to inform other modules

> about the updates to the maintained Bitcoin best chain:

Reviewer 2: ```suggestion
about the updates to the maintained Bitcoin light client:
```
